### PR TITLE
New API to pull ADIF QSO data for a defined station

### PR DESCRIPTION
**Pull request by DB4SCW**
Added a new API endpoint to pull ADIF QSO data out of wavelog from an external software using a read API key.

This enables other software to consistently pull new emerging QSO data for a station without wavelog having to deal with the "accounting" specifics. This will allow smaller software to integrate with wavelog without requiring wavelog to explicitly support each and every niche one of them.

Full disclosure: This API would be used by Hamawardz.app, a award tracking software written by me. Not yet open source, but it will be. This API would be an alternative to the native manual ADIF upload for the users.

Testing can be done like that:
`curl -d '{"key":"MYKEY","station_id":1, "fetchfromid":1}' http://log.address.here/index.php/api/get_contacts_adif`

### CODE REVIEW

File: application/controllers/Api.php [hunk 1]
```diff
@@ -237,6 +237,101 @@ function qso($dryrun = false) {
...
+		{
+			http_response_code(400);
+			echo json_encode(['status' => 'failed', 'reason' => "Not all required fields were present in input JSON"]);
+			return;
+		}
+
+		//extract relevant data to variables
+		$key = $obj['key'];
+		$station_id = $obj['station_id'];
+		$goalpost = $obj['goalpost'];
```
[the remark this answers is not in the document]
Comment (HB9HIL) [reply application/controllers/Api.php:274]: I did not dig into this yet but I'm a friend of clear and intuitive wording. :)

**Review by int2001 — APPROVED** [review]
i like the idea behind the new API-Endpoint.
From security and functional-aspects this is well done.

i'm going to approve it, but let's wait at least for one more collaborator's approval until merge.

**Comment by HB9HIL** [comment]
Hi @DB4SCW 

I took a look and I really like your project and this Pull Request. There is just one thing which is currently not clear to me. 

You created a new file `exportapi.php` with the ADIF header for your json. Why do you not use the existing `exportall.php` as the output is exactly the same? For me it looks like redundant code but maybe I miss something here.

**Comment by DB4SCW** [comment]
Hey  @HB9HIL,

I could not reuse exportall.php because of this line:
`header('Content-Disposition: attachment; filename="'.$this->session->userdata('user_callsign').'-'.date('Ymd-Hi').'.adi"')`

This forced a download of the adif file, even though this line should just render the view to variable:
`$adif_content = $this->load->view('adif/data/exportapi', $data, TRUE);`

I was contemplating adding conditions to those statements, but I did not want to change anything in the core ADIF creation view, as that would have meant changing every call to this view supplying some kind of variable to control the behaviour.

If you like me to change that to reuse the file and add conditions to the view and every call to it, I'm happy to do it.

**Comment by HB9HIL** [comment]
> Hey @HB9HIL,
> 
> I could not reuse exportall.php because of this line: `header('Content-Disposition: attachment; filename="'.$this->session->userdata('user_callsign').'-'.date('Ymd-Hi').'.adi"')`
> 
> This forced a download of the adif file, even though this line should just render the view to variable: `$adif_content = $this->load->view('adif/data/exportapi', $data, TRUE);`
> 
> I was contemplating adding conditions to those statements, but I did not want to change anything in the core ADIF creation view, as that would have meant changing every call to this view supplying some kind of variable to control the behaviour.
> 
> If you like me to change that to reuse the file and add conditions to the view and every call to it, I'm happy to do it.

That's a good argument. Probably we have to clean up one day on our side. Until then we may leave this `exportapi.php` there. 

Last Question: Why `goalpost` :D 

As far I can see it set's the last fetched QSO ID (COL_PRIMARY_KEY) so your API won't call the whole bunch of QSO's again and again. As @int2001 already mentioned we really like clear names. 

From my side please accomplish these last ToDo's

- [ ] If you are willed, change the name of `goalpost` or add a comment at the first declaration what `goalpost` is. 
- [ ] Please add a comment in `application/controllers/Api.php` what the Api does ("JSON Cake with ADIF Whip inside")

Thank you very very much for contributing ❤️  We will keep an eye on Hamawardz.app 😃 👍

**Comment by DB4SCW** [comment]
Happy to do the cleanup in another pull request if you like me to.

I chose goalpost because it sets a baseline and anything "beyond the goalpost" is a goal and as such, fair game for export. I wanted to refrain from using just "from", because for me, "from" implies datetime, which this is not. I can see why that name is not 100% intuitive, though.

I have no problem changing that to @int2001 suggestion of FetchFromID and LastFetchedID, or I can comment what goalpost means. The decision is yours, just tell me.

Of course, I will add a comment describing what the endpoint does in that commit as well.

**Comment by DB4SCW** [comment]
@HB9HIL Tasks done.

**Comment by HB9HIL** [comment]
Change the obj in the PR description :) Just for documentation

image

**Review by HB9HIL — APPROVED** [review]
Great to have you onboard @DB4SCW 😃

**Pull request merged**